File: Library/PaintedSkin/src/main/java/org/alee/component/skin/pack/BaseThemeSkinPack.java
package org.alee.component.skin.pack;

import android.content.res.ColorStateList;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**********************************************************
 *
 * @author: MY.Liu
 * @date: 2021/2/12
 * @description: xxxx
 *
 *********************************************************/
abstract class BaseThemeSkinPack implements IThemeSkinPack {
    /**
     * 皮肤包类型
     */
    private final int mSkinPackType;
    /**
     * 皮肤包名称
     */
    private final String mSkinPackName;
    /**
     * {@link IThemeSkinPack} 托底皮肤包
     */
    private final IThemeSkinPack mUnderpinThemeSkinPack;
    /**
     * 皮肤包当前状态
     */
    private int mCurrentStatus = Status.NEW;
    /**
     * 皮肤资源提供者
     */
    private ISkinResourcesProvider mSkinResourcesProvider;

    BaseThemeSkinPack(@SkinPackType int skinPackType, String skinPackName, IThemeSkinPack underpinThemeSkinPack) {
        mSkinPackType = skinPackType;
        mSkinPackName = skinPackName;
        mUnderpinThemeSkinPack = underpinThemeSkinPack;
    }

    boolean onReady(@NonNull ISkinResourcesProvider skinResourcesProvider) {
        mSkinResourcesProvider = skinResourcesProvider;
        mCurrentStatus = Status.RUNNABLE;
        return true;
    }

    /**
     * 获取皮肤包 名称
     *
     * @return 一般是用皮肤包绝对路径
     */
    @Override
    public final String getName() {
        return mSkinPackName;
    }

    /**
     * 判断 当前主题皮肤包是否可用
     *
     * @return true: 皮肤包有效且可以使用；false ：皮肤包无效且无法使用
     */
    @Override
    public final boolean isAvailable() {
        return Status.RUNNABLE == getCurrentStatus() && null != mSkinResourcesProvider;
    }

    /**
     * 获取皮肤包类型
     *
     * @return {@link SkinPackType#TYPE_DEFAULT}{@link SkinPackType#TYPE_STANDARD}{@link SkinPackType#TYPE_CUSTOMIZE}
     */
    @Override
    public final int getSkinPackType() {
        return mSkinPackType;
    }

    /**
     * 获取当前状态
     *
     * @return {@link Status#NEW}\{@link Status#RUNNABLE}|{@link Status#DESTROYED}
     */
    @Override
    public final int getCurrentStatus() {
        return mCurrentStatus;
    }

    /**
     * 获取状态颜色集合
     *
     * @param resId 资源id
     * @return 结果
     */
    @Nullable
    @Override
    public final ColorStateList getColorStateList(int resId) {
        ColorStateList temp;
        if (isAvailable()) {
            temp = mSkinResourcesProvider.getColorStateList(resId);
            if (null != temp) {
                return temp;
            }
        }
        return null == mUnderpinThemeSkinPack ? null : mUnderpinThemeSkinPack.getColorStateList(resId);
    }

    /**
     * 获取状态颜色集合
     *
     * @param resName 资源名称
     * @return 结果
     */
    @Nullable
    @Override
    public final ColorStateList getColorStateList(@NonNull String... resName) {
        ColorStateList temp;
        if (isAvailable()) {
            temp = mSkinResourcesProvider.getColorStateList(resName);
            if (null != temp) {
                return temp;
            }
        }
        return null == mUnderpinThemeSkinPack ? null : mUnderpinThemeSkinPack.getColorStateList(resName);
    }

    /**
     * 根据id 获取颜色资源
     *
     * @param resId 资源id
     * @return 结果
     */
    @Override
    public final int getColor(int resId) {
        int color = 0;
        if (isAvailable()) {
            color = mSkinResourcesProvider.getColor(resId);
            if (0 != color) {
                return color;
            }
        }
        return null == mUnderpinThemeSkinPack ? color : mUnderpinThemeSkinPack.getColor(resId);
    }

    /**
     * 根据资源名 获取颜色值
     *
     * @param resName 资源名
     * @return 结果
     */
    @Override
    public final int getColor(String... resName) {
        int color = 0;
        if (isAvailable()) {
            color = mSkinResourcesProvider.getColor(resName);
            if (0 != color) {
                return color;
            }
        }
        return null == mUnderpinThemeSkinPack ? color : mUnderpinThemeSkinPack.getColor(resName);
    }

    /**
     * 根据id 获取可拉伸资源
     *
     * @param resId 资源id
     * @return 结果
     */
    @Nullable
    @Override
    public final Drawable getDrawable(int resId) {
        Drawable drawable;
        if (isAvailable()) {
            drawable = mSkinResourcesProvider.getDrawable(resId);
            if (null != drawable) {
                return drawable;
            }
        }
        return null == mUnderpinThemeSkinPack ? null : mUnderpinThemeSkinPack.getDrawable(resId);
    }

    /**
     * 根据名称 获取可拉伸资源
     *
     * @param resName 资源名
     * @return 结果
     */
    @Nullable
    @Override
    public final Drawable getDrawable(String... resName) {
        Drawable drawable;
        if (isAvailable()) {
            drawable = mSkinResourcesProvider.getDrawable(resName);
            if (null != drawable) {
                return drawable;
            }
        }
        return null == mUnderpinThemeSkinPack ? null : mUnderpinThemeSkinPack.getDrawable(resName);
    }

    @Nullable
    @Override
    public final Drawable getMipmap(int resId) {
        Drawable drawable;
        if (isAvailable()) {
            drawable = mSkinResourcesProvider.getMipmap(resId);
            if (null != drawable) {
                return drawable;
            }
        }
        return null == mUnderpinThemeSkinPack ? null : mUnderpinThemeSkinPack.getMipmap(resId);
    }

    @Nullable
    @Override
    public final Drawable getMipmap(String... resName) {
        Drawable drawable;
        if (isAvailable()) {
            drawable = mSkinResourcesProvider.getMipmap(resName);
            if (null != drawable) {
                return drawable;
            }
        }
        return null == mUnderpinThemeSkinPack ? null : mUnderpinThemeSkinPack.getMipmap(resName);
    }

    /**
     * 根据可拉伸资源id获取图片资源
     *
     * @param resId id
     * @return 结果
     */
    @Nullable
    @Override
    public final Bitmap getBitmapForDrawable(int resId) {
        Bitmap bitmap;
        if (isAvailable()) {
            bitmap = mSkinResourcesProvider.getBitmapForDrawable(resId);
            if (null != bitmap) {
                return bitmap;
            }
        }
        return null == mUnderpinThemeSkinPack ? null : mUnderpinThemeSkinPack.getBitmapForDrawable(resId);
    }

    /**
     * 根据可拉伸资源名称获取图片资源
     *
     * @param resName 资源名称
     * @return 结果
     */
    @Nullable
    @Override
    public final Bitmap getBitmapForDrawable(String... resName) {
        Bitmap bitmap;
        if (isAvailable()) {
            bitmap = mSkinResourcesProvider.getBitmapForDrawable(resName);
            if (null != bitmap) {
                return bitmap;
            }
        }
        return null == mUnderpinThemeSkinPack ? null : mUnderpinThemeSkinPack.getBitmapForDrawable(resName);
    }

    /**
     * 获取主题
     *
     * @return {@link Resources.Theme}
     */
    @Override
    public final Resources.Theme getTheme() {
        return mSkinResourcesProvider.getTheme();
    }
}

File: Library/PaintedSkin-Kotlin/src/main/java/org/alee/component/skin/core/pack/BaseThemeSkinPack.kt
package org.alee.component.skin.core.pack

import android.content.res.ColorStateList
import android.content.res.Resources.Theme
import android.graphics.drawable.Drawable
import org.alee.component.skin.constant.SkinPackStatus
import org.alee.component.skin.core.resources.provider.ISkinResourcesProvider
import java.util.concurrent.atomic.AtomicInteger

/**
 * 皮肤包基类
 *
 * <p> 详细描述
 *
 * @author MingYu.Liu
 * created in 2023/8/30
 *
 */
internal sealed class BaseThemeSkinPack(
    /**
     * 皮肤包名称
     */
    private val mSkinPackName: String,
    /**
     * @see IThemeSkinPack 托底皮肤包
     */
    private val mUnderpinThemeSkinPack: IThemeSkinPack?,
) : IThemeSkinPack {

    /**
     * 皮肤包当前状态
     */
    private val mCurrentStatus = AtomicInteger(SkinPackStatus.NEW)

    /**
     * @see ISkinResourcesProvider
     */
    private lateinit var mSkinResourcesProvider: ISkinResourcesProvider

    /**
     * 皮肤包就绪
     * @param provider ISkinResourcesProvider [ISkinResourcesProvider]
     * @return Boolean true:已就绪
     */
    internal fun onReady(provider: ISkinResourcesProvider): Boolean {
        mSkinResourcesProvider = provider
        mCurrentStatus.set(SkinPackStatus.RUNNABLE)
        return true
    }

    /**
     * 皮肤包名称
     */
    override val name: String
        get() = mSkinPackName

    /**
     * 判断 当前主题皮肤包是否可用
     * <p> true: 皮肤包有效且可以使用；false ：皮肤包无效且无法使用
     */
    override val isAvailable: Boolean
        get() = SkinPackStatus.RUNNABLE == state && this::mSkinResourcesProvider.isInitialized

    /**
     * 皮肤包当前状态
     */
    override val state: Int
        get() = mCurrentStatus.get()

    /**
     * 主题
     */
    override val theme: Theme?
        get() = if (isAvailable) mSkinResourcesProvider.theme else null

    /**
     * 获取状态颜色集合
     * @param resId Int 颜色资源id
     * @return ColorStateList?
     */
    override fun getColorStateList(resId: Int): ColorStateList? {
        return fetchResource { provider -> provider.getColorStateList(resId) }
    }

    /**
     * 获取状态颜色集合
     * @param resNames Array<out String> 资源名称
     * @return ColorStateList?
     */
    override fun getColorStateList(vararg resNames: String): ColorStateList? {
        return fetchResource { provider -> provider.getColorStateList(resNames = resNames) }
    }

    /**
     *  根据资源id 获取颜色资源
     * @param resId Int 资源id
     * @return Int?
     */
    override fun getColor(resId: Int): Int? {
        return fetchResource { provider -> provider.getColor(resId) }
    }

    /**
     * 根据资源名 获取颜色值
     * @param resNames Array<out String> 资源名称
     * @return Int?
     */
    override fun getColor(vararg resNames: String): Int? {
        return fetchResource { provider -> provider.getColor(resNames = resNames) }
    }

    /**
     * 根据资源id 获取可拉伸资源
     * @param resId Int 资源id
     * @return Drawable?
     */
    override fun getDrawable(resId: Int): Drawable? {
        return fetchResource { provider -> provider.getDrawable(resId) }
    }

    /**
     * 根据名称 获取可拉伸资源
     * @param resNames Array<out String> 资源名称
     * @return Drawable?
     */
    override fun getDrawable(vararg resNames: String): Drawable? {
        return fetchResource { provider -> provider.getDrawable(resNames = resNames) }
    }

    /**
     * 根据资源id 获取mipmap目录下的图片资源
     * @param resId Int 资源id
     * @return Drawable?
     */
    override fun getMipmap(resId: Int): Drawable? {
        return fetchResource { provider -> provider.getMipmap(resId) }
    }

    /**
     * 根据名称 获取mipmap目录下的图片资源
     * @param resNames Array<out String> 资源名称
     * @return Drawable?
     */
    override fun getMipmap(vararg resNames: String): Drawable? {
        return fetchResource { provider -> provider.getMipmap(resNames = resNames) }
    }

    private inline fun <T> fetchResource(block: (ISkinResourcesProvider) -> T?): T? {
        if (isAvailable) {
            block(mSkinResourcesProvider)?.run {
                return this
            }
        }
        return mUnderpinThemeSkinPack?.run(block)
    }
}
